### ISSUE TITLE
Auto Benchmark disabled but RBM ties up my rig with sometimes over 100 benchmarks anyway

### ISSUE THREAD

**Issue by smyers63** [RainbowMiner/RainbowMiner#2311]
This is really costing me money.  I've said before when I reported other benchmarking problems that I should be the one to decide when benchmarks happen.  It seemed like this was better for a while but over the past few days I've had to manually add in several algorithms to exclude because even though I turned off auto benchmarking in config.txt, RBM still started running dozens and dozens of them anyway after I changed some pool settings.  Running that many benchmarks takes hours or even a day or more to complete.  I'm getting close to shutting down RBM on all 4 rigs I have running and switching back to other mining solutions.

When RBM ignores my settings and starts all these benchmarks, you still get the dev fee but I get nothing but the electric bill because the benchmarks are mostly or all for junk coins and algorithms.  I shouldn't have to contend with this again and again and again.  I never did with the other mining software I used before starting to use RBM a few months ago.  I switched to RBM because it was easier to configure for different pools but I'm starting to regret it.  Maybe I should have spent more time figuring out how to add the pools I wanted to the other software.  I'm in this for the money and lately RBM has been taking as much or more away from me than it's bringing in.

Make me a believer in RBM again.

**Comment by RainbowMiner** [comment]
Hi, I am sorry, that this thing frustrates you. I get your problem, so let us solve it. First, I should explain the parameter "EnableAutoBenchmark" a bit more: it doesn't enable/disable the basic benchmarks for missing hashrates, but selects what to be done, if a miner gets updated. So if there are missing hashrates for some miner/pool combination (Active Miners), those will be benchmarked.
So what we do need is another setting, something like "DisableBenchmarks". Setting this to "1" would then really avoid any benchmarks, no matter what. I'll add this now.

**Comment by RainbowMiner** [comment]
> So what we do need is another setting, something like "DisableBenchmarks". Setting this to "1" would then really avoid any benchmarks, no matter what. I'll add this now.

I keep thinking, that isn't what you need, so I'll not add it. But I already have another idea in mind.

Let me explain the current, basic benchmarking:
- RainbowMiner looks at all possible algorithm/miner pairings
- if it cannot find a benchmark for one of the algo/miner pairings in the database, it will run a benchmark, no matter what
- if a miner is updated, only the hashrate stats of the updated algorithms will be deleted from the database, thus triggering a benchmark

And herein lies the problem: if you add new algorithms via pool addition or new wallets that create new algo/miner pairs and for which there are no hashrate stats in the db, a benchmark will be started. If we would now disable benchmarks completely, the new pairings will be ignored, but, on the other hand, if a miner get's an update for some algorithm, that miner would not be re-benchmarked again. RainbowMiner doesn't know, if a missing hashrate for an algo/miner pair is due to the fact, that it's been newly added through configuration, or if it's due to an miner update.

So, what I can say it, that
- there will be no more mega-benchmarks after RainbowMiner has worked through them.
- only the updated algo/miner pairs will get benchmarks, after that first benchmark
- during benchmarks, RainbowMiner always uses your wallets. There will never be a donation run.

My idea for an enhancement (instead of disabling benchmarks completely) would be to add some kind of automatically maintained top 10 best-of algorithms list for each of CPU, AMD and NVIDIA and a parameter to enable this list. What you think?

**Comment by RainbowMiner** [comment]
A second idea for an enhancement: add a button "Save current algorithms" - that would then simply fill the "Algorithms" field in config.txt with all algorithms for which there are benchmarks in the database. That would avoid addition of any new algorithms, you would need to add them in config.txt by hand.

**Comment by smyers63** [comment]
I'm still doing some experiments/tests and thinking about this.  I'll have some answers soon.

In the meantime, while doing some experiments on the problem where Luxor won't mine and there aren't balances (I'll respond to that specific issue in its existing thread shortly), I disabled all pools but the one I wanted to test and when I re-enabled those others, here come the benchmarks again.  This is happening after all the same benchmarks have already been done.  Those should not be deleted/discarded/forgotten just because one or more pools is disabled temporarily.  The same benchmarks that were there yesterday and an hour ago should still be there today so none of these should have to be done again.

So again, after disabling some pools today for only a few minutes and re-enabling them, I get this:
"WARNING: Benchmarking in progress: 114 miners left, interval is set to 60 seconds."

That should not be the intended or actual behavior.  That's twice in less than a week!

I'm glad to know that the fee isn't as high as I thought it could be during this but the miner software fees are still going to be taken, I'd think.  All the while, I'm getting nothing because most or all the coins/algos don't pay anything.  Not to mention those benchmarks were done a few days ago and at least another time before that at some point.

Other software I've used to control mining ONLY does benchmarks when I specifically instruct the software to do them and I control which ones are done and for how long.  There can, and often are, benchmarks that are months and months old.  I don't want to benchmark low-pay coins and algos more than once.  That's more or less what I'm expecting here.  We can continue to talk about how best to achieve that - and I'm not expecting you to make changes like this just because that's what I'm expecting from the software.  I'm not like that.  I just need to be making money for the greatest amount of time possible whether it's with your software or some other software.  I'm glad you're willing to try to work out some ways to help with that.

There are still some things I need to test and think about before giving you answers to the other questions but I'll do that as soon as I can.

Thanks.

**Comment by RainbowMiner** [comment]
> One other thing I thought about was that the benchmarks for a pool I had enabled either for the first time or the first time in a long while is that since I already had all the benchmarks for Zergpool, why should I have to go through all the same benchmarks again for ZergpoolSolo? ...

Clearly, you shouldn't. The pool profit values and miner/algorithm benchmarks are separate. So if it happens, something must be wrong. Let me recap, you have removed all pools but one, then re-added the other pools, then RainbowMiner did suddenly start the benchmarks again, correct?
Did you restart RainbowMiner in between, or did you just edit the config while the software was running? In fact, it shouldn't make a difference, but if I go on a bug hunt, then those things could make a difference.

**Comment by smyers63** [comment]
In that case, I wanted to test Luxor so I disabled all the pools except Luxor.  Once RBM picked up the changes, I checked out Luxor a few minutes.  Then, I re-enabled the pools that I had disabled and also added the 2 solos for Zergpool.  Zergpool and ZergpoolCoins had been enabled before and all the benchmarks for them had been done previously.  The 2 extras I enabled were ZergpoolSolo and ZergpoolCoinsSolo.  That's when new benchmarks started - I think it was 114 for the 2 solo pools I added that weren't already there before.  I did not restart or exit RBM.  I changed the config while it was running.

I would think the Zergpool benchmarks could be used for ZergpoolSolo and the ZergpoolCoins benchmarks could be used for ZergpoolCoinsSolo.  Is there a reason that wouldn't work?   And could the benchmarks for Zergpool also be used for ZergpoolCoins?  That would cut the number of benchmarks for all of Zergpool and the time they would take by 75% and would be a huge improvement.  Hopefully that can be done with Zergpool and also with any other pools that have a pool and a solo option.

I'll try to remember to backup the benchmarks before I make changes like that again but it's hard to always remember.  Often, I make some change and commit to it just before realizing I had intended to do something else first.

Yes, having a button is a good reminder that it's possible.
